src/main/java/uk/ac/cam/alpha2017/airqualityradar/backend/data/HistoricalDataProvider.java

```diff
@@ -29,15 +28,60 @@ public HistoricalDataProvider(AzureTableConnector connector, DataPointConverter
      * each of the specified date/times, at each of the specified locations
      *
      * @param fromCalendar The date and time to start return data points from
-     * @param toCalendar The date and time that we will not return data points after
+     * @param toCalendar   The date and time that we will not return data points after
      * @return The data points from the historical data per calendar per location
      */
     public List<DataPoint> getDataPoints(Calendar fromCalendar, Calendar toCalendar) throws StorageException, InvalidKeyException, URISyntaxException, TableDoesNotExistException {
-        Iterator<AirDataEntity> resultIterator = connector.getEntitiesBetweenCalendars(TABLE_NAME, fromCalendar, toCalendar);
-        List<DataPoint> dataPoints = new ArrayList<>();
+        ArrayList<AirDataEntity> airDataResultList = new ArrayList<>();
+        ArrayList<WeatherDataEntity> weatherDataResultList = new ArrayList<>();
 
-        resultIterator.forEachRemaining(entity -> dataPoints.add(dataPointConverter.convertAirAndWeatherToDataPoint(entity, null)));
+        // Convert iterators to list using Java8 syntactic sugar & lambdas <3
+        connector.getEntitiesBetweenCalendars(HISTORICAL_AIR_DATA_TABLE_NAME, fromCalendar, toCalendar, AirDataEntity.class).forEachRemaining(airDataResultList::add);
+        connector.getEntitiesBetweenCalendars(HISTORICAL_WEATHER_DATA_TABLE_NAME, fromCalendar, toCalendar, WeatherDataEntity.class).forEachRemaining(weatherDataResultList::add);
 
+        // Will error if no weatherDataResults retrieved but I think that's appropriate (otherwise mapDataPoints would error)
```

Comment:
I can see where you're coming from, but in the case that we don't find any weather points, I think that it would perhaps make more sense to still return DataPoints, as there is still air quality information.  We could just set the weatherDataPoint to null.  What do you think?